src/libcore/option.rs

```diff
@@ -633,6 +633,117 @@ impl<T> Option<T> {
     }
 
     /////////////////////////////////////////////////////////////////////////
+    // Entry-like operations to insert if None and return a reference
+    /////////////////////////////////////////////////////////////////////////
+
+    /// Inserts `v` into the option if it is None, then returns a reference to
+    /// the contained value
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(option_entry)]
+    ///
+    /// let mut x = None;
+    /// assert_eq!(x.get_or_insert(5), &5);
+    /// assert_eq!(x.get_or_insert(6), &5);
+    /// ```
+    #[inline]
+    #[unstable(feature = "option_entry", issue = "39288")]
+    pub fn get_or_insert(&mut self, v: T) -> &T {
+        match *self {
+            None => *self = Some(v),
+            _ => (),
+        }
+
+        match *self {
+            Some(ref v) => v,
+            _ => unreachable!(),
+        }
+    }
+
+
+    /// Inserts a value computed from `f` into the option if it is None, then
+    /// returns a reference to the contained value
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(option_entry)]
+    ///
+    /// let mut x = None;
+    /// assert_eq!(x.get_or_insert_with(|| 5), &5);
+    /// assert_eq!(x.get_or_insert_with(|| 6), &5);
+    /// ```
+    #[inline]
+    #[unstable(feature = "option_entry", issue = "39288")]
+    pub fn get_or_insert_with<F: FnOnce() -> T>(&mut self, f: F) -> &T {
+        match *self {
+            None => *self = Some(f()),
+            _ => (),
+        }
+
+        match *self {
+            Some(ref v) => v,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Inserts `v` into the option if it is None, then
+    /// returns a mutable reference to the contained value
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(option_entry)]
+    ///
+    /// let mut x = None;
+    /// let y: &mut u32 = x.get_mut_or_insert(5);
+    /// assert_eq!(y, &5);
+    /// *y = 7;
```

Comment:
I didn't do it on purpose, because the only thing that was interesting here was the mutability of the returned value and I wanted the example shorter. But I am happy to add it if you think that's better?